src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp

```diff
@@ -365,11 +375,25 @@ inline oop ShenandoahHeap::try_evacuate_object(oop p, Thread* thread, Shenandoah
       switch (target_gen) {
         case YOUNG_GENERATION: {
            copy = allocate_from_gclab(thread, size);
+           if ((copy == nullptr) && (size < ShenandoahThreadLocalData::gclab_size(thread))) {
+             // GCLAB allocation failed because we are bumping up against the limit on young evacuation reserve.  Try resetting
+             // the desired GCLAB size and retry GCLAB allocation to avoid cascading of shared memory allocations.
+             ShenandoahThreadLocalData::set_gclab_size(thread, PLAB::min_size());
+             copy = allocate_from_gclab(thread, size);
+             // If we still get nullptr, we'll try a shared allocation below.
+           }
            break;
         }
         case OLD_GENERATION: {
            if (ShenandoahUsePLAB) {
-             copy = allocate_from_plab(thread, size);
+             copy = allocate_from_plab(thread, size, is_promotion);;
+             if ((copy == nullptr) && (size < ShenandoahThreadLocalData::plab_size(thread))) {
+               // PLAB allocation failed because we are bumping up against the limit on old evacuation reserve.  Try resetting
+               // the desired PLAB size and retry PLAB allocation to avoid cascading of shared memory allocations.
+               ShenandoahThreadLocalData::set_plab_size(thread, PLAB::min_size());
+               copy = allocate_from_gclab(thread, size);
```

Comment:
Should this be `allocate_from_plab` as above?

Reply:
Yes. Thank you for catching that.